external_search.py

```diff
@@ -882,8 +577,748 @@ def create_new_version(cls, search_client, base_index_name, version=None):
+        # call Query() recursively.
+        #
+        # However, someone who does this kind of search is
+        # probably not looking for a specific book. They might be
+        # looking for an author (eg, 'science fiction iain
+        # banks'). But they're most likely searching for a _type_
+        # of book, which means a match against summary or subject
+        # ('asteroids') would be the most useful.
+        if (self.final_query_string
+            and self.final_query_string != self.original_query_string):
+            match_rest_of_query = self.query_class.simple_query_string_query(
+                self.final_query_string,
+                self.SIMPLE_QUERY_STRING_FIELDS
+            )
+            self.match_queries.append(match_rest_of_query)
+
+    def add_match_query(self, query, field, query_string, matched_portion):
+        """Create a match query that finds documents whose value for `field`
+        matches `query`.
+
+        Add it to `self.match_queries`, and remove the relevant portion
+        of `query_string` so it doesn't get reused.
+        """
+        if not query:
+            # This is not a relevant part of the query string.
+            return query_string
+        match_query = self.query_class._match(field, query)
+        self.match_queries.append(match_query)
+        return self._without_match(query_string, matched_portion)
+
+    def add_target_age_query(self, query, query_string, matched_portion):
+        """Create a query that finds documents whose value for `target_age`
+        matches `query`.
+
+        Add it to `match_queries`, and remove the relevant portion
+        of `query_string` so it doesn't get reused.
+        """
+        if not query:
+            # This is not a relevant part of the query string.
+            return query_string
+        match_query = self.query_class.make_target_age_query(query, 40)
+        self.match_queries.append(match_query)
+        return self._without_match(query_string, matched_portion)
+
+    @classmethod
+    def _without_match(cls, query_string, match):
+        """Take the portion of a query string that matched a controlled
+        vocabulary, and remove it from the query string, so it
+        doesn't get reused later.
+        """
+        # If the match was "children" and the query string was
+        # "children's", we want to remove the "'s" as well as
+        # the match. We want to remove everything up to the
+        # next word boundary that's not an apostrophe or a
+        # dash.
+        word_boundary_pattern = r"\b%s[\w'\-]*\b"
+
+        return re.compile(
+            word_boundary_pattern % match.strip(), re.IGNORECASE
+        ).sub("", query_string)
+
+
+class Filter(SearchBase):
+    """A filter for search results.
+
+    This covers every reason you might want to not show a search
+    result that matches the query string -- wrong media, wrong
+    language, not available in the patron's library, etc.
```

Comment:
This is true but could be misleading - elasticsearch applies the filters first to reduce the number of items that have to be scored with the query, rather than finding all matches and then filtering them.